src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -707,28 +731,60 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalStartBackgroundWork(_In_ BackgroundCall
     st = pthread_attr_setdetachstate(&attrs, PTHREAD_CREATE_DETACHED);
     ASSERT(st == 0);
 
+    const char* name_copy = nullptr;
+    if (name)
+    {
+        const int MAX_THREAD_NAME_SIZE = 15;
+        char* name_copying = new (nothrow) char[MAX_THREAD_NAME_SIZE + 1];
+        if (name_copying)
+        {
+            strncpy(name_copying, name, MAX_THREAD_NAME_SIZE);
+            name_copying[MAX_THREAD_NAME_SIZE] = '\0';
+        }
+        name_copy = name_copying;
+    }
+
     pthread_t threadId;
+#ifdef __APPLE__
+    struct ThreadNameThunkContext* context = new (nothrow) ThreadNameThunkContext();
```

Comment:
It may look better to change the user thread proc to call `SetThreadName` as the first thing that it does.

It would avoid these `__APPLE__` ifdefs.

Reply:
I am trying to avoid having to change the `GCToEEInterface`. Right now the abstraction there is that we start a thread with a name. (And it is the primary motivation for this change since PerfView depended on thread names to detect ServerGC [here](https://github.com/microsoft/perfview/blob/0732e763490f296c70cb38e6857a65c21ac95d19/src/TraceEvent/Computers/TraceManagedProcess.cs#L552))

I guess I could also make the `GCToEEInterface` implementation to leave that thunk instead of inside the PAL, that can limit the thunk for just those threads.

Reply:
> I guess I could also make the GCToEEInterface implementation to leave that thunk instead of inside the PAL, that can limit the thunk for just those threads.

Yes. `GCToEEInterface::CreateThread` existing thunks for other reasons. You can pass the name in addition to what's there already.

Reply:
The new iteration should have this new design - the thread names for event pipe and finalizer are set as well.